### ISSUE TITLE
Import fails if composite user types are used

### ISSUE THREAD

**Issue by garcia556** [pgmodeler/pgmodeler#883]
There is a problem with pgmodeler while trying to import composite user-defined type that contains another user-defined type. Pgmodeler imports table with a column of top-level UDT, finds underlying non-existent UDT and raises error. That error is hidden due to reprocessing functionality (you can disable it in Diff menu but not in Import) but after that during second import of initially failed object error is getting raised regarding existing permissions (at least for my schema) as following:

    [5] src/databasemodel.cpp (2645)
      void DatabaseModel::addPermission(Permission *)
        [ERR_ASG_DUPLIC_PERMISSION] There is already a permission on object `transaction' (Table) which has one or more equal roles from those present on permission to be assigned to the object!
    [4] src/databasemodel.cpp (2671)
      void DatabaseModel::addPermission(Permission *)
        [ERR_ASG_DUPLIC_PERMISSION] There is already a permission on object `transaction' (Table) which has one or more equal roles from those present on permission to be assigned to the object!
    [3] src/databaseimporthelper.cpp (2010)
      void DatabaseImportHelper::createPermission(attribs_map &)
        [ERR_ASG_DUPLIC_PERMISSION] There is already a permission on object `transaction' (Table) which has one or more equal roles from those present on permission to be assigned to the object!
    [2] src/databaseimporthelper.cpp (502)
      void DatabaseImportHelper::createPermissions()
        [ERR_ASG_DUPLIC_PERMISSION] There is already a permission on object `transaction' (Table) which has one or more equal roles from those present on permission to be assigned to the object!
    [1] src/databaseimporthelper.cpp (641)
      void DatabaseImportHelper::importDatabase()
        [ERR_ASG_DUPLIC_PERMISSION] There is already a permission on object `transaction' (Table) which has one or more equal roles from those present on permission to be assigned to the object!
    [0] src/databaseimportform.cpp (405)
      void DatabaseImportForm::captureThreadError(Exception)
        [ERR_ASG_DUPLIC_PERMISSION] There is already a permission on object `transaction' (Table) which has one or more equal roles from those present on permission to be assigned to the object!

So to overcome it I tried to find initial exception raised, it can only be captured via debugging since exception variable is not set. It's getting raised in PgSQLTypes::operator =

```c++
	type_idx=getBaseTypeIndex(type_name);
	usr_type_idx=getUserTypeIndex(type_name, nullptr);

	if(type_idx==0 && usr_type_idx==0)
		throw Exception(ERR_ASG_INV_TYPE_OBJECT,__PRETTY_FUNCTION__,__FILE__,__LINE__);
```

Stack trace for it:

    1 PgSQLType::operator=(String const&):pgsqltypes.cpp:936
    2 PgSQLType::PgSQLType(String const&):pgsqltypes.cpp:736
    3 PgSQLType::parseString(String const&):pgsqltypes.cpp:856
    4 DatabaseImportHelper::createType(std::map<QString, QString>&):databaseimporthelper.cpp:1480
    5 DatabaseImportHelper::createObject(std::map<QString, QString>&):databaseimporthelper.cpp:711
    6 DatabaseImportHelper::getDependencyObject(QString const&, ...):databaseimporthelper.cpp:810
    7 DatabaseImportHelper::createTable(std::map<QString, QString>&):databaseimporthelper.cpp:1603
    8 DatabaseImportHelper::createObject(std::map<QString, QString>&):databaseimporthelper.cpp:712
    9 DatabaseImportHelper::createObjects():databaseimporthelper.cpp:320

I see the following solutions:
1) import UDT first recursively going downstream traversing the types hierarchy
2) create any missing UDT while importing tables

Example how to reproduce this: let's say we have three user-defined types:
- ut_base (data int)
- ut_base_ext (base ut_base, data_ext int)
- ut_base_ext_ext (base ut_base_ext, data_ext_ext int)

One is enveloped into another, *ut_base* <- *ut_base_ext* <- *ut_base_ext_ext*.
There is a table with ut_base_ext_ext column. If you try to import this table into pgmodeler visual schema it will try to import ut_base_ext as ut_base_ext_ext subtype first and fail since it's not loaded yet.

**Comment by rkhaotix** [comment]
Hi @garcia556 

I tried to follow your procudures to reproduce the problem but pgModeler imported the model correctly without a single error. 

You can use the following model:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!--
CAUTION: Do not modify this file unless you know what you are doing.
         Unexpected results may occur if the code is changed deliberately.
-->
<dbmodel pgmodeler-ver="0.9.0-alpha1" last-position="0,0" last-zoom="1"
	 default-schema="public" default-owner="postgres">
<database name="composite_test">
</database>

<schema name="public" fill-color="#e1e1e1" sql-disabled="true">
</schema>

<usertype name="udt_base" configuration="composite">
	<schema name="public"/>
	<role name="postgres"/>
	<typeattrib name="attr">
		<type name="smallint" length="0"/>
	</typeattrib>
</usertype>

<usertype name="udt_ext" configuration="composite">
	<schema name="public"/>
	<role name="postgres"/>
	<typeattrib name="base">
		<type name="public.udt_base" length="0"/>
	</typeattrib>
	<typeattrib name="date_ext">
		<type name="smallint" length="0"/>
	</typeattrib>
</usertype>

<usertype name="udt_ext_2" configuration="composite">
	<schema name="public"/>
	<role name="postgres"/>
	<typeattrib name="attr_ext">
		<type name="public.udt_ext" length="0"/>
	</typeattrib>
	<typeattrib name="attr_int">
		<type name="smallint" length="0"/>
	</typeattrib>
</usertype>

<table name="table_test">
	<schema name="public"/>
	<role name="postgres"/>
	<position x="391" y="334"/>
	<column name="col_test">
		<type name="public.udt_ext_2" length="0"/>
	</column>
</table>

</dbmodel>
```
Just copy the code above and save in a file, for instance, ```composite_import.dbm``` and open in pgModeler. After load the model try to perform  the export and import operations and you'll see that no error is raised.

If you have a model file that can trigger the bug please send it to me.

Thanks!

**Comment by garcia556** [comment]
Hey @rkhaotix,

This is something related to my own schema that I'm importing from the database to pgmodeler. I tried to isolate the table with three types and I'm still receiving the error. However after I tried to export it via pg_dump (with --oids or without) and created database from scratch applying dump the issue was gone. Probably there is something going on with OIDs that are changed during dump/restore process.

The one thing I could do is to restore a filesystem backup on remote server so you could try importing model from the database directly.

Here are connection settings for pgmodeler:
![2016-12-09_01-04-50](https://cloud.githubusercontent.com/assets/3230176/21039413/bec0b7a4-bdab-11e6-807b-50c32f69fd45.png)

Please go to "Import" form, choose connection/database "drc_pgmod", hit "Import" and you'll get the following error:
![2016-12-09_01-07-56](https://cloud.githubusercontent.com/assets/3230176/21039445/f5715678-bdab-11e6-8bb6-3e2c399a02c5.png)

Thanks!

**Comment by rkhaotix** [comment]
Nice! I will try to import from the server you indicated.

Thanks

Em 9 de dez de 2016 04:10, "garcia556" <notifications@github.com> escreveu:

**Comment by rkhaotix** [comment]
Hi @garcia556 

Good news! The problem was easier to solve than I thought.

pgModeler was wrongly duplicating permissions of some objects when running the recreation mechanism. A simple checking fixed the bug and I was able the import the database without errors :)

You can test the fix in this commit 2db0c2b.

**Comment by garcia556** [comment]
Thanks @rkhaotix, it's working now! Closing this off.
